javascript/ql/src/Security/CWE-798/HardcodedCredentials.ql

```diff
@@ -16,30 +16,50 @@
 import javascript
 import semmle.javascript.security.dataflow.HardcodedCredentialsQuery
 import DataFlow::PathGraph
+import semmle.javascript.filters.ClassifyFiles
 
 bindingset[s]
 predicate looksLikeATemplate(string s) { s.regexpMatch(".*((\\{\\{.*\\}\\})|(<.*>)|(\\(.*\\))).*") }
 
+predicate updateMessageWithSourceValue(string value, DataFlow::Node source, DataFlow::Node sink) {
+  exists(string val | val = source.getStringValue() |
+    // exclude dummy passwords and templates
+    not (
+      sink.(Sink).(DefaultCredentialsSink).getKind() = ["password", "credentials", "token", "key"] and
+      PasswordHeuristics::isDummyPassword(val)
+      or
+      sink.(Sink).getKind() = "authorization header" and
+      PasswordHeuristics::isDummyAuthHeader(val)
+      or
+      looksLikeATemplate(val)
+    ) and
+    value = "The hard-coded value \"" + val + "\""
+  )
+}
+
 from Configuration cfg, DataFlow::PathNode source, DataFlow::PathNode sink, string value
 where
   cfg.hasFlowPath(source, sink) and
-  // use source value in message if it's available
-  if source.getNode().asExpr() instanceof ConstantString
+  // sink kind is "jwt key" and source is constant string
+  if
+    sink.getNode().(Sink).(DefaultCredentialsSink).getKind() = "jwt key" and
+    // use source value in message if it's available
+    source.getNode().asExpr() instanceof ConstantString
   then
-    exists(string val | val = source.getNode().getStringValue() |
-      // exclude dummy passwords and templates
-      not (
-        sink.getNode().(Sink).(DefaultCredentialsSink).getKind() =
-          ["password", "credentials", "token", "key"] and
-        PasswordHeuristics::isDummyPassword(val)
-        or
-        sink.getNode().(Sink).getKind() = "authorization header" and
-        PasswordHeuristics::isDummyAuthHeader(val)
-        or
-        looksLikeATemplate(val)
-      ) and
-      value = "The hard-coded value \"" + val + "\""
-    )
-  else value = "This hard-coded value"
+    not isTestFile(sink.getNode().getFile()) and
+    updateMessageWithSourceValue(value, source.getNode(), sink.getNode())
+  else
+    // sink kind is "jwt key" and source is not constant string
+    if
+      sink.getNode().(Sink).(DefaultCredentialsSink).getKind() = "jwt key" and
+      not source.getNode().asExpr() instanceof ConstantString
+    then not isTestFile(sink.getNode().getFile()) and value = "This hard-coded value"
+    else
+      // sink kind is not "jwt key" and source is  constant string
+      if
+        not sink.getNode().(Sink).(DefaultCredentialsSink).getKind() = "jwt key" and
+        source.getNode().asExpr() instanceof ConstantString
+      then updateMessageWithSourceValue(value, source.getNode(), sink.getNode())
+      else value = "This hard-coded value"
```

Comment:
Can you revert these changes, and then instead put the filter on the `Sink` class instead?

Reply:
I'll have a look this week. I just came back from a few weeks of vacation.

Reply:
You can e.g. do it like this: https://github.com/erik-krogh/ql/commit/38e5c90d2719d090a9c42c46bbf4e3f8f9c6ab17 
(I tried to push directly to your PR, but I can't do that).  

There should probably also be a comment on the new characteristic predicate in `DefaultCredentialsSink`.